src/Language/Fixpoint/Solver/PLE.hs

```diff
@@ -266,9 +271,97 @@ withAssms :: InstEnv a -> ICtx -> Diff -> Maybe SubcId -> (InstEnv a -> ICtx ->
+
+  let smtRewrites = if etabeta ieCfg then
+                      concatMap (\i -> extractRewrites $ lookupBindEnv i ieBEnv) delta
+                    else
+                      mempty
+
+  SMT.smtBracket ieSMT "PLE.evaluate" $ do
+    forM_ assms (SMT.smtAssert ieSMT )
+    act env' ctx' { icAssms = mempty, icSMTRewrites = M.union (M.fromList smtRewrites) (icSMTRewrites ctx) }
+  where
+
+    -- [(is$Something foo);
+    --  (~((is$SomethingElse foo));
+    --  ...
+    --  ((prop foo) = 
+    --    (Something v₁ ... vₙ));
+    --  ((prop foo) =
+    --    (Something m₁ ... mₙ));
+    --  ...]
+
+    -- We solve the system v₁ = m₁, v₂ = m₂, ... vₙ = mₙ
+    -- finding equalities between the free variables
+
+    -- Given ((Something v₁ ... vₙ), (Something m₁ ... mₙ))
+    -- or    (v₁,                    (Something m₁ ... mₙ))
+    -- obtain the unification
+    obtainEqualities (e1, e2)
+      | fst1 : args1 <- flatten $ dropECst e1
+      , fst2 : args2 <- flatten $ dropECst e2
+      , EVar cons1 <- dropECst fst1
+      , EVar cons2 <- dropECst fst2
+      , cons1 == cons2
+      , Just _ <- M.lookup cons1 (knDataCtors ieKnowl)
+      = unifyArgumets args1 args2
+    obtainEqualities (e1, e2)
+      | EVar f <- dropECst e1
+      , validVar f
+      , fst2 : _ <- flatten $ dropECst e2
+      , (EVar cons) <- dropECst fst2
+      , Just _ <- M.lookup cons $ knDataCtors ieKnowl
+      = Just [(f, e2)]
+    obtainEqualities (e1, e2)
+      | EVar f <- dropECst e2
+      , validVar f
+      , fst1 : _ <- flatten $ dropECst e1
+      , (EVar cons) <- dropECst fst1
+      , Just _ <- M.lookup cons $ knDataCtors ieKnowl
+      = Just [(f, e1)]
+    obtainEqualities _ = Nothing
+
+    flatten (EApp x xs) = flatten x ++ [ xs ]
+    flatten x = [x]
+
+    unifyArgumets [] [] = Just []
+    unifyArgumets (x:xs) (y:ys)
+      | EVar xn <- dropECst x
+      , validVar xn
+      = fmap ((xn, y) :) (unifyArgumets xs ys)
+    unifyArgumets (x:xs) (y:ys)
+      | EVar yn <- dropECst y
+      , validVar yn
+      = fmap ((yn, x) :) (unifyArgumets xs ys)
+    unifyArgumets (x:xs) (y:ys)
+      | dropECst x == dropECst y
+      = unifyArgumets xs ys
+    unifyArgumets _ _ 
+      = Nothing
+
+    extractRewrites (_, sreft, _)
+      | Reft (_, PAnd ra) <- sr_reft sreft
+      =  concat
+      $  Mb.mapMaybe obtainEqualities
+      $  Mb.mapMaybe (keepTrans . snd)
+      $  Misc.groupList 
+      $  bimap unApply unApply
+     <$> Mb.mapMaybe keepPropEq ra
+    extractRewrites _ = []
+
+    keepPropEq (PAtom Eq e1 e2) = Just (e1, e2)
+    keepPropEq _ = Nothing
+    
+    keepTrans [a, b] = Just (a, b)
+    keepTrans _ = Nothing
+
+-- Find a way to make this check more roboust
+-- Check if a symbol is a local vairable
+validVar :: Symbol -> Bool
+validVar name =  not (T.elem '.' name')
+              && (T.elem '#' name' || T.elem '_' name')
+              && not (T.isPrefixOf "lq_anf" name')
+  where name' = symbolText name
```

Comment:
Local in the sense that the argument of a function ex:

```haskell
{-@ f :: a:Int -> b:Int -> { a + b = b + a } @-}
```

`a` and `b` are local

Reply:
I see. But I'm afraid that it is a concept from liquidhaskell, liquid-fixpoint has no notion of function specification, does it?

We have instead an environment of bindings, subtyping constraints, and function definitions. Could we explain what PLE is supposed to do, if we lack the concept of function specs?

Reply:
The core difference is that top level definitions types ecc have a definition, "locals" don't have one so you can infer their definition during unification

Reply:
As I see it, these equalities do not define the variables. Instead, they assert additional assumptions about them, whether the variable is local doesn't seem to invalidate them.

Do you have an example where it goes wrong for non-local variables?

Reply:
You can remove the check and try SKIEta.hs

Reply:
Thanks, though that's going to be some debugging exercise. If you already understand the interaction, you could try putting your account of it somewhere for the sake of the review.

Reply:
Yeah in a comment of here? The intuition is imaginenyou have to unify `local_variable = real_function`
You would like only the rewrite `local_variable` -> `real_function` and not the opposite as other than being fairly useless it would lead to an infinite loop

Reply:
> Yeah in a comment of here?

Eventually in a comment, but either works for now.

> and not the opposite as other than being fairly useless it would lead to an infinite loop

That's helpful to know. Which leads me back to figuring out why PLE needs to apply these rewrites to begin with before continuing.

Reply:
Yeah lets take as an example this [theorem](https://github.com/AlecsFerra/liquidhaskell/blob/29ee6e1a524b195bfb97bcd0a9550f497aba0c2a/tests/ple/pos/SKIEta.hs#L239) here knowing what f is equal to allows us to perform more unfoldings and prove more things, let's focus on this [case](https://github.com/AlecsFerra/liquidhaskell/blob/29ee6e1a524b195bfb97bcd0a9550f497aba0c2a/tests/ple/pos/SKIEta.hs#L262) in this specific case by unification we obtain that `f = lookupRef ... (Here ...)` so PLE now will unfold `lookupRef` and so on... genertaing the various qualities (as we have a specific instance/call of `lookupRef`), instead if we don't perform the rewrite yes the SMT will still know that `f = lookupRef ... (Here ...)` but it won't unfold the call to lookup red and subsequently it won't know all the other equalities

Reply:
if you know that `f = lookupRef σ (Cons σ γ) (Here ...)` then you perform the substitution in the equation and you get
```
\eta1 -> VFun σ σ (lookupRef σ (Cons σ γ) (Here ...) (With σ γ x eta1))
==> // Definition of lookupRef
VFun σ σ (\eta1 -> \eta1)
```
The alpha renaming in not the job of PLE, it's done for free by the SMT

Do you think that [this](https://github.com/AlecsFerra/liquidhaskell/blob/29ee6e1a524b195bfb97bcd0a9550f497aba0c2a/tests/ple/pos/ReaderEta.hs#L46) is better? I cited the SKI example because it was the one that started everything

Reply:
> if you know that f = lookupRef σ (Cons σ γ) (Here ...)

Oh, I got it now. I updated the proof. Will think about it some more.

> Do you think that this is better?

Maybe, we would have to elaborate the proof to see.

Reply:
Here's a trimmed down example that could work for explanation, together with references to your tests:
```
definitions
g x y = y
id x = x

Environment:
v = C g
v = C f

Constraint:
D (f x) ==> D id

Proof:

D (f x)
==>   // unification: (C g = C f ==> f = g)
D (g x)
==>   // eta expansion
D (\eta1 -> g x eta1)
==>   // definition of g
D (\eta1 -> eta1)
==>   // definition of id
D id
```

Reply:
Deferred by removal of the change.